elements/lisk-chain/src/state_store/state_store.ts

```diff
@@ -60,9 +63,34 @@ export class StateStore {
 		this.chain.restoreSnapshot();
 	}
 
-	public finalize(batch: BatchChain): void {
-		this.account.finalize(batch);
-		this.chain.finalize(batch);
-		this.consensus.finalize(batch);
+	public finalize(height: string, batch: BatchChain): void {
+		const accountStateDiff = this.account.finalize(batch);
+		const chainStateDiff = this.chain.finalize(batch);
+		const consensusStateDiff = this.consensus.finalize(batch);
+		this._saveDiff(
+			height,
+			[accountStateDiff, chainStateDiff, consensusStateDiff],
+			batch,
+		);
+	}
+
+	// eslint-disable-next-line class-methods-use-this
+	private _saveDiff(
+		height: string,
+		stateDiffs: Array<Readonly<StateDiff>>,
+		batch: BatchChain,
+	): void {
+		const diffToEncode = stateDiffs.reduce(
+			(acc, val) => {
+				acc.updated.push(...val.updated);
+				acc.created.push(...val.created);
+				return acc;
+			},
+			{ updated: [], created: [] },
+		);
+
+		codec.addSchema(stateDiffSchema);
```

Comment:
this `addSchema` should only happen once. It should be moved to somewhere that is doing other schema